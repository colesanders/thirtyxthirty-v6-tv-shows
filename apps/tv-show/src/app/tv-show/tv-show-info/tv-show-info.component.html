<mat-card class="mat-card-dark"> 
    <mat-card-header style="display: flex; justify-content: space-between;">
      <mat-card-title class="accent">
        <span *ngIf="tvshowForm.value.id !== null; else elseBlock">
          Editing TvShow
        </span>
        <ng-template #elseBlock>
          Create TvShow
        </ng-template>
      </mat-card-title>

      <span>
        <span *ngFor="let star of starDisplay | slice:0:tvshowForm.value.stars">
          <mat-icon svgIcon="thirty-star" class="star-color"></mat-icon>
        </span>
      </span>
      
    </mat-card-header>
  <form [formGroup]="tvshowForm" (ngSubmit)="saved.emit()">
    <mat-card-content >

      <mat-form-field class="full-width" style="width: 50%" >
        <mat-label *ngIf="tvshowForm.invalid && tvshowForm.touched">
          Name must contain at least 2 characters.
        </mat-label>
        <input  matInput required placeholder="Name" type="string"
        formControlName="name">
      </mat-form-field>
      <br>
    
      <mat-form-field appearance="fill">
        <mat-label>Select Type</mat-label>
        <mat-select matInput id="type" formControlName="type">
          <mat-option value="sit-com">Sit-Com</mat-option>
          <mat-option value="rom-com">Rom-Com</mat-option>
          <mat-option value="kids">Kids</mat-option>
          <mat-option value="teen">Teen</mat-option>
          <mat-option value="action-drama">Action-Drama</mat-option>
        </mat-select>
      </mat-form-field>
      <br>

        

      <mat-form-field class="full-width">
        <mat-label *ngIf="tvshowForm.invalid && tvshowForm.touched">
          Description must contain 10 characters.
        </mat-label>

        <textarea  matInput required placeholder="Description" type="text"
        formControlName="description">
        </textarea>
      </mat-form-field>
      <br>

      <mat-form-field class="full-width">
        <mat-label>
          {{tvshowForm.value.price | currency}}
        </mat-label>
          <input  matInput required placeholder="Price" type="number"
          formControlName="price">
      </mat-form-field> 
      <br>
      
      <mat-radio-group matInput required class="content-spread" formControlName="stars">
        <mat-radio-button [value]="1">1 Star</mat-radio-button>
        <mat-radio-button [value]="2">2 Stars</mat-radio-button>
        <mat-radio-button [value]="3">3 Stars</mat-radio-button>
        <mat-radio-button [value]="4">4 Stars</mat-radio-button>
        <mat-radio-button [value]="5">5 Stars</mat-radio-button>
      </mat-radio-group>

    </mat-card-content>
    <mat-card-actions style="display: flex; justify-content: space-between;">
      <button [disabled]="tvshowForm.invalid" type="submit" mat-button color="primary">Save</button>
      <button type="button" mat-button (click)="cancelled.emit();tvshowForm.reset();">
        Cancel
      </button>
    </mat-card-actions>
  </form>
</mat-card>