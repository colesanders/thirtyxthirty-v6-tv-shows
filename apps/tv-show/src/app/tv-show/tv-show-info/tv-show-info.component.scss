@import '~@angular/material/theming';

mat-form-field {
    width: 100%;
}

mat-slider {
    width: 100%;
    margin: auto;
}

.star-color {
    color: gold;
}

h4 {
    text-align: center; 
    color: mediumblue;
}

mat-icon {
    padding-top: 8px;
}

mat-radio-button {
    padding-right: 5px;
}