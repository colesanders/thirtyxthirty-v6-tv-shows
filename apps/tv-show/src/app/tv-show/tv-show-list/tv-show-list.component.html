<mat-card class="mat-card-dark">
    <mat-card-header>
      <mat-card-title class="accent">
        TvShows List
      </mat-card-title>
    </mat-card-header>
    <mat-card-content>
      <mat-list>
        <mat-list-item *ngFor="let tvshow of tvshows" (click)="selected.emit(tvshow)" style="cursor: pointer">
          <h3 matLine> {{ tvshow.name | titlecase }} </h3>
          <p matLine>
            {{ tvshow.description }}
          </p>
          <button mat-icon-button color="warn" (click)="deleted.emit(tvshow); $event.stopImmediatePropagation()" >
            <mat-icon>delete</mat-icon>
          </button>
        </mat-list-item>
      </mat-list>
      
    </mat-card-content>
</mat-card>