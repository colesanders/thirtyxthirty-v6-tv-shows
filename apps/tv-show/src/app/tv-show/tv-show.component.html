<div class="tvshow-containers">

    <div class="tvshow-list">
        <thirty-tv-show-list [tvshows]="tvshow$ | async" (selected)="selectTvShow($event)" (deleted)="deleteTvShow($event)">
        </thirty-tv-show-list>
    </div>
    <div class="tvshow-info">
        <thirty-tv-show-info [tvshowForm]="tvshowForm" (saved)="saveTvShow()" (cancelled)="resetForm()"> 
        </thirty-tv-show-info>
    </div>
</div>
