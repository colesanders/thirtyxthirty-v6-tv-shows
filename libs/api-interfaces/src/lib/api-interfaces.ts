export interface Message {
  message: string;
}

export interface Fruit {
  "id": number,
  "name": string,
  "description": string,
  "color": string,
  "favorite": boolean,
  "icon" : string,
  "amount": number,
}

export interface TvShow {
  "id": number,
  "name": string,
  "description": string,
  "type": string,
  "price": number,
  "stars": number,
}

export interface Game {
  "id": number,
  "name": string,
  "description": string,
  "rating": string,
  "price": number,
}

export interface Car {
  "id": number,
  "type": string,
  "description": string,
  "color": string,
  "price": number,
}

export interface Login {
  "id": number,
  "username": string,
  "password": string,
  "logged_in": boolean,
}
