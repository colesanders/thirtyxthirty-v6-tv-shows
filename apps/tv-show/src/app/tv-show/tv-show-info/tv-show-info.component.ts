import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { TvShow } from '@thirty/api-interfaces';
import {FormGroup } from '@angular/forms';

@Component({
  selector: 'thirty-tv-show-info',
  templateUrl: './tv-show-info.component.html',
  styleUrls: ['./tv-show-info.component.scss']
})
export class TvShowInfoComponent implements OnInit {
  @Input() tvshowForm;
  @Output() saved = new EventEmitter();
  @Output() cancelled = new EventEmitter();

  starDisplay: number[];

  constructor() { }

  ngOnInit(): void {
    this.starDisplay = [1,1,1,1,1];
  }
}
