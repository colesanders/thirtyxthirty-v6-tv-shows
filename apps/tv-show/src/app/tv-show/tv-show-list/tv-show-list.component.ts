import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { TvShow } from '@thirty/api-interfaces';

@Component({
  selector: 'thirty-tv-show-list',
  templateUrl: './tv-show-list.component.html',
  styleUrls: ['./tv-show-list.component.scss']
})
export class TvShowListComponent implements OnInit {
  @Input() tvshows: [TvShow];
  @Output() selected = new EventEmitter<TvShow>();
  @Output() deleted = new EventEmitter<TvShow>();
  constructor() { }

  ngOnInit(): void {
  }

}
