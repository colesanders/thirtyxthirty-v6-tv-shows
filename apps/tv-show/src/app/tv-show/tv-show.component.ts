import { Component, OnInit } from '@angular/core';
import { TvShow } from '@thirty/api-interfaces';
import { TvShowService, SnackBarService } from '@thirty/core-data';
import { Observable } from 'rxjs';
import { FormGroup, FormBuilder, FormControl, Validators } from '@angular/forms';

@Component({
  selector: 'thirty-tv-show',
  templateUrl: './tv-show.component.html',
  styleUrls: ['./tv-show.component.scss']
})
export class TvShowComponent implements OnInit {
  title = 'TvShow Store';
  selectedTvShow: TvShow;
  tvshowForm: FormGroup
  tvshow$: Observable<TvShow[]>;

  constructor(private tvshowService: TvShowService, private formBuilder: FormBuilder, public snackBarService: SnackBarService) { }

  ngOnInit(): void {
    this.creatFormGroup();
    this.refresh();
  }

  deleteTvShow(tvshow:TvShow) {
    this.tvshowService.deleteTvShow(tvshow)
    .subscribe( m => {
      this.refresh();
    });
  }

  saveTvShow() {
    const tvshow = this.tvshowForm.value;

    this.snackBarService.openSnackBar("TvShow Saved!", "Okay", 2000)

    if(tvshow.id !== null){
      this.tvshowService.updateTvShow(tvshow)
      .subscribe( m => {
        this.refresh();
      });
    } else {
      this.tvshowService.createTvShow(tvshow)
      .subscribe( m => {
        this.refresh();
      });
    }
  }

  refresh(){
    this.loadTvShows();
    this.resetForm();
  }

  creatFormGroup(){
    this.tvshowForm = this.formBuilder.group({
      id: [],

      name: new FormControl('', [
        Validators.required,
        Validators.minLength(2),
      ]),
      description: new FormControl('', [
        Validators.required,
        Validators.minLength(10),
      ]),
      type: new FormControl('', [
        Validators.required,
      ]),
      price: new FormControl(0, [
        Validators.required,
      ]),
      stars: new FormControl(1, [
        Validators.required,
      ]),
    })
  }

  resetForm() {
    this.tvshowForm.patchValue({});
  }

  selectTvShow(tvShow: TvShow) {
    this.tvshowForm.patchValue(tvShow);
  }

  loadTvShows(): void{
    this.tvshow$ = this.tvshowService.getTvShows();
  }

}
