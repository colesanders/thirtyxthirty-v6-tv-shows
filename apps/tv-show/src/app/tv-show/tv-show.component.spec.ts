import { HttpClientModule } from '@angular/common/http';
import { DebugElement } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { TvShowService } from '@thirty/core-data';
import { MaterialModule } from '@thirty/material';
import { of } from 'rxjs';
import { TvShowComponent } from './tv-show.component';
import { TvShow } from '@thirty/api-interfaces';
import { TvShowInfoComponent } from './tv-show-info/tv-show-info.component';
import { TvShowListComponent } from './tv-show-list/tv-show-list.component';

const mockEmptyTvShow: TvShow = {
  id: null,
  name: 'Mock Empty TvShow',
  description: '',
  type: '',
  price: 0,
  stars: 1,
};

const mockTvShow = {
  id: 1,
  name: 'Mock TvShow',
  description: 'Mock descr',
  type: 'sit-com',
  price: 0,
  stars: 1,
};

const mockTvShowService = {
  getTvShows: () => of([]),
  createTvShow: () => of({}),
  updateTvShow: () => of({}),
  deleteTvShow: () => of({})
}

describe('TvShowComponent', () => {
  let component: TvShowComponent;
  let fixture: ComponentFixture<TvShowComponent>;
  let de: DebugElement;
  let tvShowService: TvShowService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        ReactiveFormsModule,
        HttpClientModule,
        MaterialModule,
        NoopAnimationsModule,
        RouterTestingModule,
      ],
      providers: [
        { provide: TvShowService, useValue: mockTvShowService }
      ],
      declarations: [
        TvShowComponent,
        TvShowInfoComponent,
        TvShowListComponent
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TvShowComponent);
    component = fixture.componentInstance;
    de = fixture.debugElement;
    tvShowService = de.injector.get(TvShowService);
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should select a tvShow', () => {
    component.selectTvShow(mockTvShow);
    expect(component.tvshowForm.value).toMatchSnapshot();
  });
});
