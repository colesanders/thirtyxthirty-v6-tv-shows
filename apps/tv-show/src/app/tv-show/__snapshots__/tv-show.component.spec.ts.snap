// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TvShowComponent should select a tvShow 1`] = `
Object {
  "description": "Mock descr",
  "id": 1,
  "name": "Mock TvShow",
  "price": 0,
  "stars": 1,
  "type": "sit-com",
}
`;
